src/main/asciidoc/_chapters/offheap_read_write.adoc

```diff
@@ -0,0 +1,146 @@
+to manage the offheap memory. Similarly, on the read path, we’ll try to read the BucketCache firstly, if the Cache
+misses, go to the HFile and read the corresponding block. The workflow: reading blocks from cache OR sending cells to
+client,  is basically not involved in heap memory allocations.
+
+image::offheap-overview.png[]
+
+
+[[regionserver.offheap.readpath]]
+== Offheap read-path
+In HBase-2.0.0, link:https://issues.apache.org/jira/browse/HBASE-11425[HBASE-11425] changed the HBase read path so it
+could hold the read-data off-heap avoiding copying of cached data on to the java heap.
+This reduces GC pauses given there is less garbage made and so less to clear. The off-heap read path has a performance
+that is similar/better to that of the on-heap LRU cache.  This feature is available since HBase 2.0.0.
+If the BucketCache is in `file` mode, fetching will always be slower compared to the native on-heap LruBlockCache.
+Refer to below blogs for more details and test results on off heaped read path
+link:https://blogs.apache.org/hbase/entry/offheaping_the_read_path_in[Offheaping the Read Path in Apache HBase: Part 1 of 2]
+and link:https://blogs.apache.org/hbase/entry/offheap-read-path-in-production[Offheap Read-Path in Production - The Alibaba story]
+
+For an end-to-end off-heaped read-path, first of all there should be an off-heap backed <<offheap.blockcache>>(BC). Configure 'hbase.bucketcache.ioengine' to off-heap in
+_hbase-site.xml_. Also specify the total capacity of the BC using `hbase.bucketcache.size` config. Please remember to adjust value of 'HBASE_OFFHEAPSIZE' in
+_hbase-env.sh_. This is how we specify the max possible off-heap memory allocation for the
+RegionServer java process. This should be bigger than the off-heap BC size. Please keep in mind that there is no default for `hbase.bucketcache.ioengine`
+which means the BC is turned OFF by default (See <<direct.memory>>).
+
+Next thing to tune is the ByteBuffer pool on the RPC server side.
+The buffers from this pool will be used to accumulate the cell bytes and create a result cell block to send back to the client side.
+`hbase.ipc.server.reservoir.enabled` can be used to turn this pool ON or OFF. By default this pool is ON and available. HBase will create off heap ByteBuffers
+and pool them. Please make sure not to turn this OFF if you want end-to-end off-heaping in read path.
+If this pool is turned off, the server will create temp buffers on heap to accumulate the cell bytes and make a result cell block. This can impact the GC on a highly read loaded server.
+The user can tune this pool with respect to how many buffers are in the pool and what should be the size of each ByteBuffer.
+Use the config `hbase.ipc.server.reservoir.initial.buffer.size` to tune each of the buffer sizes. Default is 64 KB.
+
+When the read pattern is a random row read load and each of the rows are smaller in size compared to this 64 KB, try reducing this.
```

Comment:
Not a comment on this patch.
Now we have the same BB pool to be used for reading the HFile blocks as well as for sending back the RPC responses.  For smaller row sizes we say here that users can tune the BB size to be lower. But that might impact the read from HFiles. I assume we don't change the block sizes in HFiles but keep the 64KB default.  So is this suggest us to have a large as well as small sized pool? Just an open Q for thinking.

Reply:
Yeah, If  user have a very small cell size, such as 100B, then shouldn't tune the buffer.size to 1KB or some thing smaller,  because of the block size.   Keep the 65KB  will be more better. So this  suggestion can be removed now I think.  Say even if keep a cell with 100B in a 65KB buffer, it won't have much impact, the buffer will put back to the pool and sending a cell  is quite short time.